### PR TITLE
KAFKA-7367: Ensure stateless topologies don't require disk access

### PR DESCRIPTION
**Pull request by vvcephei**
Streams should not eagerly create the state store directory, since in general topologies may not need state stores.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vvcephei** [comment]
I've currently just implemented the test (which fails). Fix pending...

The culprit is this, in the KafkaStreams constructor:
```java
        try {
            stateDirectory = new StateDirectory(config, time);
        } catch (final ProcessorStateException fatal) {
            throw new StreamsException(fatal);
        }
```

File: streams/src/test/java/org/apache/kafka/streams/integration/StatelessIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+            waitUntilMinRecordsReceived(
+                consumerConfig,
+                outputTopic,
+                1
+            );
+
+            if (new File(path).exists()) {
+                final List<Path> collect =
+                    Files.find(Paths.get(path), 999, (p, bfa) -> true).collect(Collectors.toList());
```
Comment (enothereska) [inline streams/src/test/java/org/apache/kafka/streams/integration/StatelessIntegrationTest.java:98]: This might not be robust enough if the streams app decides to store the directory elsewhere right? I'm thinking we need another test as part of this file that checks a stateful topology and confirms the storage is actually there. Then the pair of tests would be more robust to changes.

File: streams/src/test/java/org/apache/kafka/streams/integration/StatelessIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, broker.bootstrapServers());
+        streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, path);
+        streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 1000);
+        streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        streamsConfiguration.put(IntegrationTestUtils.INTERNAL_LEAVE_GROUP_ON_CLOSE, true);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+        final KStream<String, String> input = builder.stream(inputTopic, Consumed.with(Serdes.String(), Serdes.String()));
+        final KStream<String, String> filter = input.filter((k, s) -> s.length() % 2 == 0);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/StatelessIntegrationTest.java:65]: How about also augment the built `Topology` with processor APIs to add more processors, and check that as long as we do not call addStores the resulted topology would still be stateless.

**Review by guozhangwang — COMMENTED** [review]
@vvcephei Thanks for bringing this up. I think a good solution would be deferring the state directory until the tasks has been assigned (in the stream thread main loop's onPartitionAssinged callback). We need to be careful if there are any other edge cases though that may require the state directory before that event.

File: streams/src/test/java/org/apache/kafka/streams/integration/StatelessIntegrationTest.java (hunk 1 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/StatelessIntegrationTest.java:98]: Hey @enothereska , thanks for the review!

Just to make sure I follow, you're saying that we should have a test that verifies that Streams actually respects its state directory config?

I agree that it would be a serious bug if it did not, so if we're lacking test coverage on that, I'd be happy to add it.

File: streams/src/test/java/org/apache/kafka/streams/integration/StatelessIntegrationTest.java (hunk 2 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/StatelessIntegrationTest.java:65]: Ok, yes, that would be a good additional test.

File: streams/src/test/java/org/apache/kafka/streams/integration/StatelessIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            waitUntilMinRecordsReceived(
+                consumerConfig,
+                outputTopic,
+                1
+            );
+
+            if (new File(path).exists()) {
+                final List<Path> collect =
+                    Files.find(Paths.get(path), 999, (p, bfa) -> true).collect(Collectors.toList());
+                Assert.fail(path + " should not have existed, but it did: " + collect);
```
Comment (kamalcph) [inline streams/src/test/java/org/apache/kafka/streams/integration/StatelessIntegrationTest.java:99]: `TestUtils.tempDirectory(applicationId).getPath()` - creates the temp directory. The `collect` variable includes the application directory. So, the assert fail always.

Is it allowed to create `application` directory for stateless topology ?

Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/StatelessIntegrationTest.java:99]: heh. This is a really good point. I'll update the test to use a subdirectory of the temp directory as the state directory, and then assert that the state directory doesn't exist.

It is currently allowed for Streams to auto-create the state directory even if it doesn't need to. I wrote KAFKA-7367 because I don't think it should, though, and this test is intended to be the spec to verify that any implementation of lazy state dir creation actually doesn't automatically create the directory.

**Comment by vvcephei** [comment]
No need to keep this PR open, it was only ever to demonstrate an issue.
